### ISSUE TITLE
Transition Mosaic effect is slightly inaccurate

### ISSUE THREAD

**Issue by fmatthew5876** [EasyRPG/Player#2068]
Do frame grab of Mosaic transition effect and you'll see our algo is slightly off in how it chooses which pixels to blow up. Not enough for players to really notice, but it could be fixed.

I haven't looked into this deeply so I don't have more details other than that.

**Comment by Ghabry** [comment]
Found no time yet. I will take a look at the tables next week.

**Comment by Ghabry** [comment]
actually the offsets into the mosaic seem to be random. The 2nd I took screenshots different offsets are choosen.

**Comment by Ghabry** [comment]
(This post doesn't include the random offset as the algorithm for this is unclear)

I also wasn't able to understand your table and found a simpler pattern:

The pattern I observed now is (starting with scale 2) `-1, -1, -2, -2, -3, -3, -4, -4`, so `-scale / 2` (integer division).

Where the number is the initial (X,Y) coordinate where the algorithm begins. So for x2 scale it is (-1, -1). From their it fills a scale-sized square down-right with that colour (nearest neighbour), this repeats until the entire image is filled.

When the reference pixel (top-left of the square) is out of bounds it takes the pixel closest (edge handling = extend, see the image).

![Image](https://github.com/user-attachments/assets/4adf9e09-afbe-4120-9460-f97933477c6d)

(Source: https://commons.wikimedia.org/wiki/File:Extend_Edge-Handling.png)

-----

Here a ZIP with again a kra and a psd file to look at the result (when you compare it with the files in mosaic.zip: They match).

Also included a jupyter notebook file which I used to figure out the algorithm.

[Mosaic2.zip](https://github.com/user-attachments/files/21008356/Mosaic2.zip)

**Comment by Ghabry** [comment]
These 2 inner loops that fill the rectangle are cache/prefetch unfriendly as they make the drawing operation not go row-by-row anymore.

Considering this is always "pick top left pixel in the square" and this sounds like a very common problem when doing nearest neighbour I asked AI to do the math for me:

```python3
def clamp(x, lower, upper):
    return max(lower, min(x, upper))

def s(scale):
    # The offset defines where at (X,Y) the pixel is picked for scaling (nearest neighbour)
    # The pixel is usually initially out of bounds
    # in this case the nearest pixel of the image is choosen (edge handling = extend)
    imgo = img.copy()
    w = 320
    h = 240

    off = -(scale // 2)

    for row in range(h):
        # Compute which input pixel this row belongs to
        src_row = ((row - off) // scale) * scale + off
        src_row = clamp(src_row, 0, h - 1)

        for col in range(w):
            src_col = ((col - off) // scale) * scale + off
            src_col = clamp(src_col, 0, w - 1)

            imgo[row][col] = img[src_row][src_col]

    return imgo
```

**Comment by Ghabry** [comment]
Yeah and I traced the random numbers: they are in range [0, scale).
